examples/vault-iam-auth/main.tf

```diff
@@ -0,0 +1,285 @@
+# ---------------------------------------------------------------------------------------------------------------------
+resource "aws_iam_instance_profile" "example_instance_profile" {
+  path = "/"
+  role = "${aws_iam_role.example_instance_role.name}"
+}
+
+resource "aws_iam_role" "example_instance_role" {
+  name_prefix        = "${var.auth_server_name}-role"
+  assume_role_policy = "${data.aws_iam_policy_document.example_instance_role.json}"
+}
+
+data "aws_iam_policy_document" "example_instance_role" {
+  statement {
+    effect  = "Allow"
+    actions = ["sts:AssumeRole"]
+
+    principals {
+      type        = "Service"
+      identifiers = ["ec2.amazonaws.com"]
+    }
+  }
+}
+
+# Adds policies necessary for running consul
+module "consul_iam_policies_for_client" {
+  source = "github.com/hashicorp/terraform-aws-consul.git//modules/consul-iam-policies?ref=v0.3.10"
+
+  iam_role_id = "${aws_iam_role.example_instance_role.id}"
+}
+
+# ---------------------------------------------------------------------------------------------------------------------
+# THE USER DATA SCRIPT THAT WILL RUN ON THE INSTANCE
+# This script will run consul, which is used for discovering vault cluster
+# And perform the login operation
+# ---------------------------------------------------------------------------------------------------------------------
+
+data "template_file" "user_data_auth_client" {
+  template = "${file("${path.module}/user-data-auth-client.sh")}"
+
+  vars {
+    consul_cluster_tag_key   = "${var.consul_cluster_tag_key}"
+    consul_cluster_tag_value = "${var.consul_cluster_name}"
+    example_role_name        = "${var.example_role_name}"
+  }
+}
+
+# ---------------------------------------------------------------------------------------------------------------------
+# ADDS A RULE TO OPEN PORT 8080 SINCE OUR EXAMPLE LAUNCHES A SIMPLE WEB SERVER
+# ---------------------------------------------------------------------------------------------------------------------
+
+resource "aws_security_group" "auth_instance" {
+  name        = "${var.auth_server_name}"
+  description = "Security group for ${var.auth_server_name}"
+  vpc_id      = "${data.aws_vpc.default.id}"
+}
+
+resource "aws_security_group_rule" "allow_inbound_api" {
+  type        = "ingress"
+  from_port   = "8080"
+  to_port     = "8080"
+  protocol    = "tcp"
+  cidr_blocks = ["0.0.0.0/0"]
+
+  security_group_id = "${aws_security_group.auth_instance.id}"
+}
+
+# ---------------------------------------------------------------------------------------------------------------------
+# ADDS A POLICY TO THE VAULT CLUSTER ROLE SO VAULT CAN QUERY AWS IAM USERS AND ROLES
+# ---------------------------------------------------------------------------------------------------------------------
+
+resource "aws_iam_role_policy" "vault_iam" {
+  name   = "vault_iam"
+  role   = "${module.vault_cluster.iam_role_id}"
+  policy = "${data.aws_iam_policy_document.vault_iam.json}"
+}
```

Comment:
Does it make sense to create a dedicated module called `vault-aws-auth` which includes these permissions? Or perhaps one module for `vault-aws-auth-iam` and one for `vault-aws-auth-ec2` if the permissions are different? It seems like an essential aspect of setting up Vault when using the AWS auth method.

Reply:
Why an independent module? If we are moving this out of here, it might as well just be an additional parameter for `vault-cluster`

Reply:
As per discussion on our zoom meeting, we came to the conlusion it would be nice to have this as a separate module. But neither of us feel strongly about this. For this reason we created an additional basecamp to-do categorised under nice-to-have